Proofread tut - add a pallet (#194)
substrate-developer-hub/substrate-docs: v3/tutorials/07-add-a-pallet/index.mdx

---

> title: Add Nicks Pallet to the Runtime

Author: ```suggestion
title: Add the Nicks Pallet to your Runtime
```

> title: Add FRAME's Nicks Pallet to Runtime

Reviewer: Either "your runtime" or "the runtime" is fine.
I think I would drop FRAME's from the title. In general, we wouldn't use the possessive like this and I am not sure we are consistent about using FRAME elsewhere.

> title: Add FRAME's Nicks Pallet to Runtime

Author: I will use **Add Nicks Pallet to the Runtime** then. I prefer *the Runtime* over *your Runtime*. *your Runtime* sounds too possessive & personal :)

> title: Add FRAME's Nicks Pallet to Runtime

Author: Updated to **Add Nicks Pallet to the Runtime** in the three files.

This guide will show you how you can add the
[Nicks pallet](/rustdocs/latest/pallet_nicks/index.html). You can follow

> similar pattern to add additional FRAME pallets to your runtime. However you should note that each

Reviewer: +1

pallet is a little different in terms of the specific configuration settings needed to use it
correctly. This tutorial will help you understand what you'll need to consider when adding a new
pallet to your FRAME runtime.

<Message
  type={`green`}
  title={`We're here to help`}

>   text={`If you run into issues on this tutorial, **we are here to help!** You can

Reviewer: +1

  [ask a question on Stack Overflow](https://stackoverflow.com/questions/tagged/substrate) and use the
  \`substrate\` tag or contact us on [Element](https://matrix.to/#/#substrate-technical:matrix.org).`}
/>

> You should already have the [latest version of the Node Template](https://github.com/substrate-developer-hub/substrate-node-template/tree/latest) compiled on your computer from when you completed the

Reviewer: ```suggestion
You should already have the [latest version of the Node Template](https://github.com/substrate-developer-hub/substrate-node-template/tree/latest) compiled on your computer if you completed the
```

[Create Your First Substrate Chain](/tutorials/v3/create-your-first-substrate-chain) tutorial. If
you do not, please complete it first.

<Message
  type={`yellow`}
  title={`Information`}
  text=

>       "Experienced developers who truly prefer to skip that tutorial may install the Node Template

Reviewer: ```suggestion
      "If you skipped the first blockchain tutorial, you should install the node template now.
```
I don't think there's any value in qualifying developers as experienced (the relevant detail is whether they did the tutorial or did not). No need for the adverb "truly" and "may" implies permission. In this case, I think we are simply saying that the node template is a prerequisite for this tutorial to work as written (and could be in a Before you begin section).

<TutorialObjective
  data={{
    textLineOne:
      '3. Use the Front-end Template to interact with the Nicks pallet.',
    url: '/tutorials/v3/add-a-pallet#interact-with-the-nicks-pallet',
  }}
/>
<TutorialObjective
  data={{
    textLineOne:
      '4. Learn how to publish a custom pallet to Github or crates.io.',
    url: '/tutorials/v3/add-a-pallet#publish-your-own-pallet',
  }}
/>

> ## Learning Outcomes

Reviewer: +1 (caps aren't user-friendly)

### Importing a Pallet Crate

The first thing you need to do is to import the [`pallet-nicks`](/rustdocs/latest/pallet_nicks) crate in your
runtime's `Cargo.toml` file. If you want a proper primer into Cargo syntax, please check out
[its official documentation](https://doc.rust-lang.org/cargo/reference/index.html).

Open `substrate-node-template/runtime/Cargo.toml` and you will see all the dependencies
your runtime has. For example, it depends on the [Balances pallet](/rustdocs/latest/pallet_balances/index.html):

**`runtime/Cargo.toml`**

```TOML
#--snip--
[dependencies.pallet-balances]
default-features = false
git = 'https://github.com/paritytech/substrate.git'

> tag = 'monthly-2021-08'

Reviewer 2: Is this the version tested with? We should do an end to end test on the latest `monthly-2021-10` per #192

Author: I think this is fine. Because in the current context, we are showing an example of what would be shown in `Cargo.toml`. We are not asking people to use this code in the tutorial.

### Importing a Pallet Crate

The first thing you need to do is to import the [`pallet-nicks`](/rustdocs/latest/pallet_nicks) crate in your
runtime's `Cargo.toml` file. If you want a proper primer into Cargo syntax, please check out
[its official documentation](https://doc.rust-lang.org/cargo/reference/index.html).

> Open `substrate-node-template/runtime/Cargo.toml` and you will see all the dependencies

Reviewer: runtime dependencies.
(Probably don't need the possessive)

from the [Nicks pallet](/rustdocs/latest/pallet_nicks/index.html). 

### Compile and run the node

Compile the node in release mode with:

```bash
cargo build --release
```

If the build fails, go back to the previous section and make sure you followed all the steps

> correctly. You can also view a [full solution including the `Nick` pallet](https://github.com/substrate-developer-hub/substrate-node-template/tree/tutorials/solutions%2Fadd-a-pallet-v3).

Reviewer 2: Looks like this is tested with latest monthly?

Author: I merge `monthlyly-2021-10` of node template into add a pallet. Refer to:
https://github.com/substrate-developer-hub/substrate-node-template/commits/tutorials/solutions/add-a-pallet-v3

This afternoon I pushed two tags `tut-sols/add-a-pallet/latest` and `tut-sols/add-a-pallet/monthly-2021-10` in node template. The pattern will be similar to the monthly tag and `latest` tag used on the main branch.

### Start the Front-End

As in the previous tutorials, this tutorial will use the

> [Substrate Front-End Template](https://github.com/substrate-developer-hub/substrate-front-end-template)

Reviewer 2: Use the monthly tagged template for consistency? Anything older than latest is incompatible AFAIK

Author: done.

refer to the [Substrate runtime](https://github.com/paritytech/substrate/tree/master/bin/node/runtime)
which includes nearly every pallets in the library of core FRAME pallets.

In the Substrate node runtime
[`bin/node/runtime/Cargo.toml` file](https://github.com/paritytech/substrate/blob/master/bin/node/runtime/Cargo.toml),

> you will see how to import different pallets, and in the

Reviewer 2: Use monthly or latest tag might want to have a new tag we float on latest template for substrate itself, like `devhub-latest`

Reviewer: When I tried the monthly tag, the tutorial wasn't successful. I was able to run through it all using the latest tag but I started with the tutorial in the old site, so it could have just been a mismatch between what I downloaded initially and when trying to re-do the tutorial using the instructions in the new substrate-docs repo.

Author: @NukeManDan I pushed a `devhub/latest` tag to Substrate repo. So ideally when pointing to rust docs, it would be `/rustdoc/latest/...` which really is the latest monthly, and when pointing to the src code it would be `substrate/tree/devhub/latest/...` which is the src code of latest monthly.

But I would suggest to keep it consistent and fix this across the whole docs after docs.substrate.io launch.

Author: @lsgunnlsgunn I re-compiled https://github.com/substrate-developer-hub/substrate-node-template/tree/tut-sols/add-a-pallet%2Flatest and found the issue. Fixed in the content (the `usize` -> `u32` data type fix) and the [code in node template](https://github.com/substrate-developer-hub/substrate-node-template/commit/b4bba5df0db6b4cff089ad021bdd5ee32d1ca670). Thanks.

### Option 2: Publishing on crates.io

> **crates.io** allows permissionless publishing of any Rust modules. You could learn the procedure by

Reviewer: ```suggestion
**crates.io** allows permissionless publishing of Rust modules. You could learn the procedure by
```
any Rust module (singluar) or leave out any and use Rust modules (plural)
